Fix memory leak in exception handling (Issue #8882)

jaiganeshs21 opened the pull request:
## Summary

This PR fixes a critical memory leak in Celery's exception handling that was causing significant memory growth when tasks raise unhandled exceptions.

## Problem

**Issue #8882** reported that Celery tasks raising unhandled exceptions were causing memory leaks:
- Memory not being garbage collected due to reference cycles
- Worse on Python 3.11+ due to enhanced traceback data

Closes #8882

This PR addresses a critical memory leak in Celery's exception handling by breaking reference cycles in tracebacks, ensuring frame locals are cleared, and adding comprehensive tests to verify the fix.

-     Add traceback_clear calls and explicit deletion of traceback references in retry, failure, and signal-handling paths

-     Enhance traceback_clear to efficiently clear frame data and update its docstring

-     Introduce unit and integration tests to validate that tracebacks no longer retain memory across various exception scenarios

auvipy commented on celery/app/trace.py:
```diff
@@ -270,6 +286,12 @@ def _log_error(self, task, req, einfo):
 
 
 def traceback_clear(exc=None):
```
@jaiganeshs21 what is your thought regarding this suggestion?

jaiganeshs21 replied:
The variable `tb` is guaranteed to be assigned before the loop in the conditional at [line 298](https://github.com/celery/celery/pull/9799/files#diff-fb194522c454315cd87ccb1cf4a2a53aafa49124c85abc01dcca4310cf19d6d3R298).

If exc has a traceback, we assign it to `tb`; otherwise, we fall back to `sys.exc_info()`. So the loop always has a valid `tb` value.

auvipy replied:
is it possible to increase unit test coverage of this changes as well?

jaiganeshs21 commented on celery/app/trace.py:
```diff
@@ -456,18 +480,22 @@ def trace_task(uuid, args, kwargs, request=None):
                     state, retval = I.state, I.retval
                     I.handle_reject(task, task_request)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
```
Same as this comment https://github.com/celery/celery/pull/9799#discussion_r2187078854:

The `traceback_clear` function does use the `exc` argument correctly.  ([traceback_clear fucntion](https://github.com/celery/celery/pull/9799/files#diff-fb194522c454315cd87ccb1cf4a2a53aafa49124c85abc01dcca4310cf19d6d3R298))

When exc is provided, it checks `hasattr(exc, '__traceback__')` and uses `exc.__traceback_`_ if available, otherwise falls back to `sys.exc_info()`.

jaiganeshs21 commented on t/integration/test_memory_leak_8882.py:
```diff
@@ -0,0 +1,247 @@
+
+
+def get_memory_usage():
+    """Get current memory usage in bytes."""
+    try:
+        import psutil
+        process = psutil.Process(os.getpid())
+        return process.memory_info().rss
+    except ImportError:
+        # Fallback to tracemalloc if psutil not available
+        current, peak = tracemalloc.get_traced_memory()
```
Got it, Explicitly documenting this behavior.

jaiganeshs21 commented:
Thanks @auvipy for adding the Copilot review and providing feedback. I've addressed the comments. Let me know if you have any other suggestions. Thanks!

auvipy commented:
btw unit tests are separated from integration tests

jaiganeshs21 commented on celery/app/trace.py:
```diff
@@ -283,8 +305,10 @@ def traceback_clear(exc=None):
 
     while tb is not None:
```
The code already has` tb = tb.tb_next` at [line 315](https://github.com/jaiganeshs21/celery/blob/fix-memory-leak-issue-8882/celery/app/trace.py#L315), at the end of the while loop. The loop does advance to the next traceback frame and terminates correctly when tb becomes None. 

Also a test was added to demonstrate this works properly - [Code](https://github.com/jaiganeshs21/celery/blob/fix-memory-leak-issue-8882/t/integration/test_mem_leak_in_exception_handling.py#L139) - process multiple frames (3/3 in our test)

jaiganeshs21 commented:
> btw unit tests are separated from integration tests

Got it, Separated the unit and integration tests and moved them to their respective locations.
Commit - https://github.com/celery/celery/pull/9799/commits/f2e24b3db474ea8b236be18f8c486de37010f720

jaiganeshs21 commented on celery/app/trace.py:
```diff
@@ -456,18 +482,22 @@ def trace_task(uuid, args, kwargs, request=None):
                     I, R = Info(REJECTED, exc), ExceptionInfo(internal=True)
                     state, retval = I.state, I.retval
                     I.handle_reject(task, task_request)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
                 except Ignore as exc:
                     I, R = Info(IGNORED, exc), ExceptionInfo(internal=True)
                     state, retval = I.state, I.retval
                     I.handle_ignore(task, task_request)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
                 except Retry as exc:
                     I, R, state, retval = on_error(
                         task_request, exc, RETRY, call_errbacks=False)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
                 except Exception as exc:
                     I, R, state, retval = on_error(task_request, exc)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
```
There is no code duplication. The current implementation is similar to what Copilot suggested.

The traceback_clear function was defined and is called in the exception handler.

We have now added a comment in every place where this function is called to explain that it is used to prevent memory retention, making it easier for code contributors to understand.

auvipy replied:
they make a lot of wrong suggestion

jaiganeshs21 commented on celery/app/trace.py:
```diff
@@ -456,18 +482,22 @@ def trace_task(uuid, args, kwargs, request=None):
                     state, retval = I.state, I.retval
                     I.handle_reject(task, task_request)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
                 except Ignore as exc:
                     I, R = Info(IGNORED, exc), ExceptionInfo(internal=True)
                     state, retval = I.state, I.retval
                     I.handle_ignore(task, task_request)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
                 except Retry as exc:
                     I, R, state, retval = on_error(
                         task_request, exc, RETRY, call_errbacks=False)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
                 except Exception as exc:
                     I, R, state, retval = on_error(task_request, exc)
+                    # MEMORY LEAK FIX: Clear traceback frames to prevent memory retention (Issue #8882)
                     traceback_clear(exc)
```
There is no code duplication. The current implementation is similar to what Copilot suggested.

The traceback_clear function was defined and is called in the exception handler.

We have now added a comment in every place where this function is called to explain that it is used to prevent memory retention, making it easier for code contributors to understand.

auvipy reviewed (commented):
overall looks good

auvipy commented on celery/app/trace.py:
```diff
@@ -270,6 +288,12 @@ def _log_error(self, task, req, einfo):
 
 
 def traceback_clear(exc=None):
+    """Clear traceback frames to prevent memory leaks.
```
@jaiganeshs21

auvipy replied:
-     Calling traceback_clear() clears the references within the traceback objects, but the exception’s __traceback__ attribute may still reference the root traceback object, which can still keep objects alive until the exception itself is collected.

-     Setting exc.__traceback__ = None after clearing ensures the exception no longer holds any references to the traceback chain.
